api/src/main/java/io/opentelemetry/trace/TraceState.java

```diff
@@ -235,26 +235,53 @@ public static Entry create(String key, String value) {
   // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
   // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
   // forward slashes /.  For multi-tenant vendor scenarios, an at sign (@) can be used to prefix the
-  // vendor name.
+  // vendor name. The tenant id (before the '@') is limited to 240 characters and the vendor id is
+  // limited to 13 characters. If in the multi-tenant vendor format, then the first character
+  // may additionally be digit.
+  //
   // todo: benchmark this implementation
   private static boolean validateKey(String key) {
-    if (key.length() > KEY_MAX_SIZE || key.isEmpty() || !isNumberOrDigit(key.charAt(0))) {
+    if (key.length() > KEY_MAX_SIZE || key.isEmpty() || !isLowercaseLetterOrDigit(key.charAt(0))) {
       return false;
     }
     int atSeenCount = 0;
     for (int i = 1; i < key.length(); i++) {
       char c = key.charAt(i);
-      if (!isNumberOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/') {
+      if (!isLowercaseLetterOrDigit(c)
+          && c != '_'
+          && c != '-'
+          && c != '@'
+          && c != '*'
+          && c != '/') {
         return false;
       }
-      if ((c == '@') && (++atSeenCount > 1)) {
-        return false;
+      int atPosition = 0;
+      if (c == '@') {
+        atPosition = i;
+        // tenant id must be 240 characters or less
+        if (i > 240) {
+          return false;
+        }
+        atSeenCount++;
+        if (atSeenCount > 1) {
+          return false;
+        }
```

Comment:
Instead of having `atPosition` is it ok to do something like this in this branch?

```java
// remaining is vendor id, must be 13 characters or less
if (key.length() - i > 13) {
  return false;
}
```